pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -962,31 +967,171 @@ func TestHighProirotyFlushUnschedulableQLeftover(t *testing.T) {
+		Type:    v1.PodScheduled,
+		Status:  v1.ConditionFalse,
+		Reason:  v1.PodReasonUnschedulable,
+		Message: "fake scheduling failure",
 	})
 
 	// Update pod condition to midPod.
 	podutil.UpdatePodCondition(&midPod.Status, &v1.PodCondition{
-		Type:          v1.PodScheduled,
-		Status:        v1.ConditionFalse,
-		Reason:        v1.PodReasonUnschedulable,
-		Message:       "fake scheduling failure",
-		LastProbeTime: metav1.Time{Time: time.Now().Add(-1 * unschedulableQTimeInterval)},
+		Type:    v1.PodScheduled,
+		Status:  v1.ConditionFalse,
+		Reason:  v1.PodReasonUnschedulable,
+		Message: "fake scheduling failure",
 	})
 
+	q.unschedulableQ.addOrUpdate(newPodInfoNoTimestamp(&highPod))
+	q.unschedulableQ.addOrUpdate(newPodInfoNoTimestamp(&midPod))
+	q.unschedulableQ.podInfoMap[util.GetPodFullName(&highPod)].timestamp = time.Now().Add(-1 * unschedulableQTimeInterval)
+	q.unschedulableQ.podInfoMap[util.GetPodFullName(&midPod)].timestamp = time.Now().Add(-1 * unschedulableQTimeInterval)
+
 	if p, err := q.Pop(); err != nil || p != &highPod {
 		t.Errorf("Expected: %v after Pop, but got: %v", highPriorityPod.Name, p.Name)
 	}
 	if p, err := q.Pop(); err != nil || p != &midPod {
 		t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPod.Name, p.Name)
 	}
 }
+
+// TestPodTimestamp tests the operations related to podInfo.
+func TestPodTimestamp(t *testing.T) {
+	pod1 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pod-1",
+			Namespace: "ns1",
+			UID:       types.UID("tp-1"),
+		},
+		Status: v1.PodStatus{
+			NominatedNodeName: "node1",
+		},
+	}
+
+	pod2 := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pod-2",
+			Namespace: "ns2",
+			UID:       types.UID("tp-2"),
+		},
+		Status: v1.PodStatus{
+			NominatedNodeName: "node2",
+		},
+	}
+
+	pInfo1 := &podInfo{
+		pod:       pod1,
+		timestamp: util.RealClock{}.Now(),
+	}
+	pInfo2 := &podInfo{
+		pod:       pod2,
+		timestamp: util.RealClock{}.Now().Add(1 * time.Second),
+	}
+
+	var queue *PriorityQueue
+	type operation = func()
+	addPodActiveQ := func(pInfo *podInfo) operation {
+		return func() {
+			queue.lock.Lock()
+			defer queue.lock.Unlock()
+			queue.activeQ.Add(pInfo)
+		}
+	}
+	updatePodActiveQ := func(pInfo *podInfo) operation {
+		return func() {
+			queue.lock.Lock()
+			defer queue.lock.Unlock()
+			queue.activeQ.Update(pInfo)
+		}
+	}
+	addPodUnschedulableQ := func(pInfo *podInfo) operation {
+		return func() {
+			queue.lock.Lock()
+			defer queue.lock.Unlock()
+			// Update pod condition to unschedulable.
+			podutil.UpdatePodCondition(&pInfo.pod.Status, &v1.PodCondition{
+				Type:    v1.PodScheduled,
+				Status:  v1.ConditionFalse,
+				Reason:  v1.PodReasonUnschedulable,
+				Message: "fake scheduling failure",
+			})
+			queue.unschedulableQ.addOrUpdate(pInfo)
+		}
+	}
+	addPodBackoffQ := func(pInfo *podInfo) operation {
+		return func() {
+			queue.lock.Lock()
+			defer queue.lock.Unlock()
+			queue.podBackoffQ.Add(pInfo)
```

Comment:
@bsalamat  thank you~ Here I add comments to facilitate your review.  Add lock upper and here to prevent data race occuring in `https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/73700/pull-kubernetes-bazel-test/76841/`